pkg/operator/sync.go

```diff
@@ -112,13 +112,18 @@ func isCloudConfigRequired(infra *configv1.Infrastructure) bool {
 	if infra.Spec.CloudConfig.Name != "" {
 		return true
 	}
-	switch infra.Status.PlatformStatus.Type {
-	case configv1.AzurePlatformType,
-		configv1.GCPPlatformType,
-		configv1.OpenStackPlatformType,
-		configv1.VSpherePlatformType:
-		return true
-	default:
+
+	if infra.Status.PlatformStatus != nil {
```

Comment:
Great job narrowing this down!

the change in `lib/resourcemerge/machineconfig.go` LGTM but following the trails for the panic indicates that our code isn't that robust to changes and we don't have units for all the functions that access the field being potentially nil (and we're human and miss things in reviews, it happens).
The original change w/o the nil check is here https://github.com/openshift/machine-config-operator/pull/1658/commits/e7455dcb4e0150e00f78e0ae4954b73047d1bf75#diff-fa45321336db7ad1cedc28bf643a4f97R107
Then, here's another (!!!) https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go#L322 (in master, doesn't affect 4.5, but as we can notice, we aren't checking for nil there and we're missing it again)

Then we can instead notice https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/render.go#L80-L91 within `createDiscoveredControllerConfigSpec` which is the first thing we do after grabbing the infra object. In those lines we make sure that if the PlatformStatus is nil, we "adjust" it for compatibility reasons and then we store the infra object in the ControllerConfig object which makes sure we can safely check that field as we made sure it's always non-nil (there's another bug there tho, unrelated, we're modifying an object from the indexers which is a no-no).
Anyway, it seems that after grabbing the initial infra object (https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go#L228), we do run it through a compatibility fix to make sure PlatformStatus is never nil from there (https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/render.go#L80-L91), perhaps a forward-looking fix for this BZ would be to:
- grab the infra object (we do this already)

- run it through `createDiscoveredControllerConfigSpec` (we do this already)
- pass the resulting `ControllerConfig.Spec.Infra...` object in funcs like this and `getIgnitionHost`, instead of the `infra` global one, as we're sure that the infra in the CC already went through this very same nil check (we need to do this)

With the above we:

- centralize the nil check (also there's a comment telling us why it could be nil, the key thing for this BZ seems to be an upgrade all the way from 4.1 which didn't have PlatformStatus)
- adjust it so it's never nil in our functions

The above allows us to avoid adding specific nil checks for the same thing over and over and help us tomorrow when the next PR adds something that access PlatformStatus w/o checking it

Reply:
sounds good!

Reply:
Me imagining what Antonio means when he says on Twitter he's going to take some time to do some writing: Antonio sitting at his computer, sipping an espresso and writing a gripping Italian crime novel :book: 

The actual result: Antonio sitting at his computer, sipping an espresso and writing longer pull request review comments :wink:

Reply:
@cgwalters 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂 😂  it was fun to imagine myself wiring a crime novel in from of my laptop sipping an expresso tho! thanks for that Colin! 😂